### PR TITLE
python3Packages.ihatemoney: 4.1 -> 4.2

### PR DESCRIPTION
**Pull request by symphorien**
Update.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by privateseabass** [comment]
What is the package name? I'm not finding it as "ihatemoney".

**Comment by symphorien** [comment]
I fixed the title of the PR. The commit messages are correct.

**Comment by symphorien** [comment]
I updated ihatemoney to 4.2 but haven't tested enough yet.

**Comment by symphorien** [comment]
@GrahamcOfBorg build python3.pkgs.ihatemoney.tests

**Comment by symphorien** [comment]
@GrahamcOfBorg build python3.pkgs.ihatemoney.tests.ihatemoney.ihatemoney-postgresql  python3.pkgs.ihatemoney.tests.ihatemoney.ihatemoney-sqlite

**Comment by symphorien** [comment]
Looks like the PR is ready.

File: pkgs/development/python-modules/debts/default.nix
Comment (timokau) [inline pkgs/development/python-modules/debts/default.nix:12]: That's a shame. In that case we should at the very least use a fully-qualified rev instead of a shortened one.

But I kind of think "no tags (-> no automatic updates)" outweighs "no tests", so I'd rather fetch it from pypi for now. Maybe create an issue against the upstream repo, notifying them that tags would be useful :)

File: pkgs/development/python-modules/debts/default.nix
Comment (timokau) [inline pkgs/development/python-modules/debts/default.nix:19]: Even more reason to just fetch it from pypi.

File: pkgs/development/python-modules/debts/default.nix
Comment (timokau) [inline pkgs/development/python-modules/debts/default.nix:23]: Why `postInstall`? Seems like this should be `checkPhase`. Note that the python `checkPhase` is actually `installCheckPhase`, so its always run after installation.

File: pkgs/development/python-modules/sqlalchemy-i18n/default.nix
Comment (timokau) [inline pkgs/development/python-modules/sqlalchemy-i18n/default.nix:1]: The mix in styles (one per line vs long lines) irks me a bit here :smile:

File: nixos/modules/services/web-apps/ihatemoney/default.nix
Comment (timokau) [inline nixos/modules/services/web-apps/ihatemoney/default.nix:136]: Released 2 days ago! https://github.com/unbit/uwsgi/releases

Probably better to make that update first & remove this workaround, since otherwise it will likely stick around forever.

**Review by timokau — CHANGES_REQUESTED** [review]
Looks mostly good, I left some comments.
/status needs_work

**Comment by timokau** [comment]
/status needs_work
That should work in review comments too.

**Comment by timokau** [comment]
Also the testsuite fails for me in `nix-review`.

File: pkgs/development/python-modules/sqlalchemy-i18n/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/sqlalchemy-i18n/default.nix:1]: I think I started doing this, everything but python packages are included in the first line, python packages are then entered 1 per line.

Generally `stdenv, lib, fetchPypi, buildPythonPackage` are unlikely to ever change, so just have them together.

I guess it's just a personal preference

File: nixos/modules/services/web-apps/ihatemoney/default.nix
Comment (symphorien) [reply nixos/modules/services/web-apps/ihatemoney/default.nix:136]: Do you want me to update uwsgi in this PR ?

**Comment by symphorien** [comment]
Ah I can reproduce the test failure when I rebase on master.
A git bisect indicates that it's caused by 86990d60071a3be8ea25d2ad2f79a0ba096331b6     pythonPackages.wtforms: 2.1 -> 2.3.1
and apparently ihatemoney only supports 2.2.1: https://github.com/spiral-project/ihatemoney/commit/53dae96bdbfcf09f4cef56afda8a7ad567458eeb

I wonder what is the way to go from there.

File: pkgs/development/python-modules/sqlalchemy-i18n/default.nix
Comment (symphorien) [reply pkgs/development/python-modules/sqlalchemy-i18n/default.nix]: I ran nixpkgs-fmt on the file :)

File: pkgs/development/python-modules/debts/default.nix
Comment (symphorien) [reply pkgs/development/python-modules/debts/default.nix]: Oh I didn't know checkPhase was after installPhase for python packages. I reworked this.

File: pkgs/development/python-modules/debts/default.nix
Comment (symphorien) [reply pkgs/development/python-modules/debts/default.nix]: changed to full revision.

**Comment by symphorien** [comment]
upstream knows about the problem https://github.com/spiral-project/ihatemoney/issues/567

File: nixos/modules/services/web-apps/ihatemoney/default.nix
Comment (timokau) [reply nixos/modules/services/web-apps/ihatemoney/default.nix:136]: Either that, or in its own PR (and ping me for review), or hope that somebody else does it soon ;)

**Comment by timokau** [comment]
Pinning dependencies in python is generally messy. For a python application we *could* get away with it, for a library we can't.

Probably should have caught this in the initial review: Is this really a library? Looks more like an application to me. Search https://nixos.org/nixpkgs/manual/ for "Building packages and applications" for more info.

Since upstream knows about it we could also
- wait for a fix there and apply the patch
- fix it for them, create a PR and *then* apply the patch.

Depending on how much work you want to invest.

**Comment by symphorien** [comment]
> Probably should have caught this in the initial review: Is this really a library? 

It is a wsgi application. That is, uwsgi starts a special python interpreter, and then *imports* ihatemoney. So for all intents and purposes, it is a library.

Regarding patching, I had a quick look, and have no idea what happens.
And regarding waiting, current version of ihatemoney is also affected.

I'll try and see if it still works if I disable this test in the coming days.

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 85023` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
<br>- python37Packages.ihatemoney
<br>- python38Packages.ihatemoney
</details>
<details>
  <summary>8 packages built:</summary>
<br>- python27Packages.sqlalchemy-continuum
<br>- python27Packages.sqlalchemy-i18n
<br>- python37Packages.debts
<br>- python37Packages.sqlalchemy-continuum
<br>- python37Packages.sqlalchemy-i18n
<br>- python38Packages.debts
<br>- python38Packages.sqlalchemy-continuum
<br>- python38Packages.sqlalchemy-i18n
</details>

**Comment by Mic92** [comment]
@GrahamcOfBorg build python37Packages.ihatemoney python38Packages.ihatemoney

**Comment by timokau** [comment]
Mic92: There is a known problem with a too-new dependency.

File: pkgs/development/python-modules/debts/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/debts/default.nix:29]: Yes this is actually a problem in `buildPythonPackage`. We should remove the folder always or change directory.

**Comment by symphorien** [comment]
I thought a bit about this, and came to the conclusion that I can override the dependencies of ihatemoney locally because ihatemoney will never be the dependency of something else. It is only a library because it must be imported by uswgi, but it will not be depended upon by some other python module.

I also updated uswgi and removed the now-unneeded workaround. I tested this in a vm, and csv downloading works :)

/status needs_review

**Comment by timokau** [comment]
I agree with your reasoning, since this is a de-facto python application. Still, since we kind of have an (unwritten?) policy against dependency overrides within python-module I'd like @FRidh's okay on this.

@FRidh do you agree that its fine to override the dependencies here, since the module will only ever be used as an application (but through uswgi)?

**Comment by timokau** [comment]
@jonringer do you maybe have an opinion on this? See my last comment.

/status needs_work
(work = feedback to get it out of the needs_review queue, will be improved by https://github.com/timokau/marvin-mk2/issues/29)

**Comment by timokau** [comment]
Well, lacking any feedback I will just assume its fine. Please add the `python.pkgs` prefix to https://github.com/NixOS/nixpkgs/pull/85023/commits/ebc28de6a9001052abdc6bd6141ce0c228275d41, then this should be good to go.

I would still prefer to avoid hardcoding the `debts` rev to enable automatic updates, but If you as the maintainer want to stay on top of updates manually, that is your call.

**Comment by symphorien** [comment]
done. The nixos tests still pass locally.

**Comment by timokau** [comment]
Thanks, and thank you for your patience!

**Pull request merged**
